packages/ai/src/ui/chat-store.ts

```diff
@@ -0,0 +1,518 @@
+    const chat = this.chats.get(id);
+    if (!chat) return;
+    return chat.messages;
+  }
+
+  getLastMessage(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+    return chat.messages[chat.messages.length - 1];
+  }
+
+  subscribe(subscriber: ChatStoreSubscriber): () => void {
+    const { id } = subscriber;
+
+    if (!this.subscribers.has(id)) {
+      this.subscribers.set(id, new Set());
+    }
+
+    this.subscribers.get(id)?.add(subscriber);
+
+    return () => this.subscribers.get(id)?.delete(subscriber);
+  }
+
+  setMessages({ id, messages }: { id: string; messages: UIMessage[] }) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+
+    chat.messages = [...messages];
+    this.emitEvent({ id, event: ChatStoreEvent.ChatMessagesChanged });
+  }
+
+  appendMessage({ id, message }: { id: string; message: UIMessage }) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+
+    if (message.role === 'user') {
+      this.resetActiveResponse(id);
+    }
+
+    chat.messages = [...chat.messages, { ...message }];
+    this.emitEvent({ id, event: ChatStoreEvent.ChatMessagesChanged });
+  }
+
+  removeAssistantResponse(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat) return;
+
+    const lastMessage = chat.messages[chat.messages.length - 1];
+
+    if (!lastMessage) {
+      throw new Error('Cannot remove assistant response from empty chat');
+    }
+
+    if (lastMessage.role !== 'assistant') {
+      throw new Error('Last message is not an assistant message');
+    }
+
+    this.setMessages({ id, messages: chat.messages.slice(0, -1) });
+    this.resetActiveResponse(id);
+  }
+
+  private emitEvent({ id, event }: { id: string; event: ChatStoreEvent }) {
+    const subscribers = this.subscribers.get(id);
+    if (!subscribers) return;
+
+    for (const subscriber of subscribers) {
+      subscriber.onChatChanged(event);
+    }
+  }
+
+  private calculateStep(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat) return 0;
+
+    if (
+      !chat.activeResponse ||
+      chat.activeResponse.message.role !== 'assistant'
+    ) {
+      return 0;
+    }
+
+    const hasToolInvocation = chat.activeResponse.message.parts?.some(
+      part => part.type === 'tool-invocation',
+    );
+
+    if (!hasToolInvocation) {
+      return 0;
+    }
+
+    return (
+      (chat.activeResponse.message.parts?.reduce((max, part) => {
+        if (part.type === 'tool-invocation') {
+          return Math.max(max, part.toolInvocation.step ?? 0) ?? 0;
+        }
+        return max;
+      }, 0) ?? 0) + 1
+    );
+  }
+
+  private resetActiveResponse(id: string) {
+    const chat = this.chats.get(id);
+    if (!chat || !chat.activeResponse) return;
+
+    chat.activeResponse = undefined;
+    chat.error = undefined;
+    chat.status = 'ready';
+    this.emitEvent({ id, event: ChatStoreEvent.ChatStatusChanged });
+    this.emitEvent({ id, event: ChatStoreEvent.ChatErrorChanged });
```

Comment:
is this separation needed or can the status change just include the error (also as additional property on the event that can undefined)?